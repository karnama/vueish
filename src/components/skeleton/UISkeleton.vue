<template>
    <div class="h-10 w-full rounded bg-gray-250 dark:bg-gray-750 text-white dark:text-gray-600"
         :style="{'--duration': `${duration}ms`}"
         :class="[pulse ? 'animate-pulse' : 'animate-shimmer']" />
</template>

<script lang="ts">
import { defineComponent } from 'vue';

export default defineComponent({
    name: 'UISkeleton',

    props: {
        /**
         * Use pulse animation instead of default shimmer.
         */
        pulse: {
            type: Boolean,
            default: false
        },

        /**
         * The duration of the animation.
         */
        duration: {
            type: [Number, String],
            default: 1300,
            validator: (num: number) => Number(num) > 0
        }
    }
});
</script>

<style lang="scss" scoped>
.animate-shimmer {
    position: relative;
    overflow: hidden;

    &::after {
        position: absolute;
        content: '';
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        transform: translateX(-100%);
        background-image: linear-gradient(
                90deg,
                theme('colors.gray.250') 0,
                theme('colors.gray.200') 20%,
                theme('colors.gray.200') 60%,
                theme('colors.gray.250')
        );
        animation-duration: var(--duration);
        animation-name: shimmer;
        animation-iteration-count: infinite;
    }

    @keyframes shimmer {
        100% {
            transform: translateX(100%);
        }
    }
}
.dark .animate-shimmer::after {
    background-image: linear-gradient(
            90deg,
            theme('colors.gray.750') 0,
            theme('colors.gray.700') 40%,
            theme('colors.gray.700') 60%,
            theme('colors.gray.750')
    );
}
.animate-pulse {
    animation-name: pulse;
    animation-duration: var(--duration);
    animation-timing-function: cubic-bezier(0.4, 0, 0.6, 1);
    animation-iteration-count: infinite;

    @keyframes pulse {
        0%, 100% {
            opacity: 1;
        }
        50% {
            opacity: .6;
        }
    }
}
</style>
