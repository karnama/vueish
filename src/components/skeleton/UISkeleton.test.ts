import { mount } from '@vue/test-utils';
import UISkeleton from 'components/skeleton/UISkeleton.vue';

describe('UISkeleton', () => {
    it('should display correctly', () => {
        let wrapper = mount(UISkeleton, {
            props: {
                pulse: false
            }
        });

        expect(wrapper.element).toMatchSnapshot();

        wrapper = mount(UISkeleton, {
            props: {
                pulse: true
            }
        });

        expect(wrapper.element).toMatchSnapshot();
    });
});
