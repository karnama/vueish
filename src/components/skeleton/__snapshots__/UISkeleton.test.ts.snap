// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UISkeleton should display correctly 1`] = `
<div
  class="h-10 w-full rounded bg-gray-250 dark:bg-gray-750 text-white dark:text-gray-600 animate-shimmer"
  style="--duration: 1300ms;"
/>
`;

exports[`UISkeleton should display correctly 2`] = `
<div
  class="h-10 w-full rounded bg-gray-250 dark:bg-gray-750 text-white dark:text-gray-600 animate-pulse"
  style="--duration: 1300ms;"
/>
`;
